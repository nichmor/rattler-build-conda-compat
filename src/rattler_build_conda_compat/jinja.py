from __future__ import annotations

from typing import Any, TypedDict

import jinja2
import yaml
from jinja2 import DebugUndefined

from rattler_build_conda_compat.loader import load_yaml


class RecipeWithContext(TypedDict, total=False):
    context: dict[str, str]


class _MissingUndefined(DebugUndefined):
    def __str__(self) -> str:
        """
        By default, `DebugUndefined` return values in the form `{{ value }}`.
        `rattler-build` has a different syntax, so we need to override this method,
        and return the value in the form `${{ value }}`.
        """
        return f"${super().__str__()}"


def jinja_env() -> jinja2.Environment:
    """
    Create a `rattler-build` specific Jinja2 environment with modified syntax.
    """

    def stub_compatible_pin(*args, **kwargs) -> str:  # noqa: ARG001, ANN003, ANN002
        return f"compatible_pin {args[0]}"

    def stub_subpackage_pin(*args, **kwargs) -> str:  # noqa: ARG001, ANN003, ANN002
        return f"subpackage_pin {args[0]}"

    def version_to_build_string(some_string: str) -> str:
        """Converts some version by removing the . character and returning only the first two elements of the version)"""
        # We first split the string by whitespace and take the first part
        split = some_string.split()[0] if some_string.split() else some_string
        # We then split the string by . and take the first two parts
        parts = split.split(".")
        major = parts[0] if len(parts) > 0 else ""
        minor = parts[1] if len(parts) > 1 else ""
        return f"{major}{minor}"

    def split_filter(s: str, sep: str = " ") -> list[str]:
        """Filter that split a string by a separator"""
        return s.split(sep)

    env = jinja2.Environment(
        variable_start_string="${{",
        variable_end_string="}}",
        trim_blocks=True,
        lstrip_blocks=True,
        autoescape=True,
        undefined=_MissingUndefined,
    )

    # inject rattler-build recipe functions in jinja environment
    env.globals.update(
        {
            "compiler": lambda x: x + "_compiler_stub",
            "stdlib": lambda x: x + "_stdlib_stub",
            "pin_subpackage": stub_subpackage_pin,
            "pin_compatible": stub_compatible_pin,
            "cdt": lambda *args, **kwargs: "cdt_stub",  # noqa: ARG005
        }
    )

    # inject rattler-build recipe filters in jinja environment
    env.filters.update(
        {
            "version_to_buildstring": version_to_build_string,
            "split": split_filter,
        }
    )
    return env


def load_recipe_context(context: dict[str, str], jinja_env: jinja2.Environment) -> dict[str, str]:
    """
    Load all string values from the context dictionary as Jinja2 templates.
    """
    # Process each key-value pair in the dictionary
    for key, value in context.items():
        # If the value is a string, render it as a template
        if isinstance(value, str):
            template = jinja_env.from_string(value)
            rendered_value = template.render(context)
            context[key] = rendered_value

    return context


def render_recipe_with_context(recipe_content: RecipeWithContext) -> dict[str, Any]:
    """
    Render the recipe using known values from context section.
    Unknown values are not evaluated and are kept as it is.

    Examples:
    ---
    ```python
    >>> from pathlib import Path
    >>> from rattler_build_conda_compat.loader import load_yaml
    >>> recipe_content = load_yaml((Path().resolve() / "tests" / "data" / "eval_recipe_using_context.yaml").read_text())
    >>> evaluated_context = render_recipe_with_context(recipe_content)
    >>> assert "my_value-${{ not_present_value }}" == evaluated_context["build"]["string"]
    >>>
    ```
    """
    env = jinja_env()
    context = recipe_content.get("context", {})
    # load all context templates
    context_templates = load_recipe_context(context, env)

    # render the rest of the document with the values from the context
    # and keep undefined expressions _as is_.
    template = env.from_string(yaml.dump(recipe_content))
    rendered_content = template.render(context_templates)
    return load_yaml(rendered_content)
