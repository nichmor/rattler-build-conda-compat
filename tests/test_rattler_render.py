from __future__ import annotations

import os
from pathlib import Path
from typing import TYPE_CHECKING, Any

from rattler_build_conda_compat.loader import parse_recipe_config_file
from rattler_build_conda_compat.render import MetaData, render

if TYPE_CHECKING:
    from pathlib import Path


def test_render_recipe(python_recipe: Path, unix_namespace: dict[str, Any], snapshot) -> None:
    variants = parse_recipe_config_file(str(python_recipe / "variants.yaml"), unix_namespace)

    rendered = render(str(python_recipe), variants=variants, platform="linux", arch="64")

    all_used_variants = [meta[0].get_used_variant() for meta in rendered]

    assert len(all_used_variants) == 2

    assert snapshot == all_used_variants


def test_environ_is_passed_to_rattler_build(env_recipe, snapshot) -> None:
    try:
        os.environ["TEST_SHOULD_BE_PASSED"] = "false"
        rendered = render(str(env_recipe), platform="linux", arch="64")
        all_used_variants = [meta[0].meta for meta in rendered]
        assert len(all_used_variants) == 1
        # for this scenario recipe should not be rendered
        assert snapshot == all_used_variants

        os.environ["TEST_SHOULD_BE_PASSED"] = "true"

        rendered = render(str(env_recipe), platform="linux", arch="64")

        all_used_variants = [meta[0].meta for meta in rendered]
        assert len(all_used_variants) == 1
        # for this scenario recipe should be rendered
        assert snapshot == all_used_variants[0]["build_configuration"]["variant"]

    finally:
        os.environ.pop("TEST_SHOULD_BE_PASSED", None)


def test_metadata_for_single_output(feedstock_dir_with_recipe: Path, rich_recipe: Path) -> None:
    (feedstock_dir_with_recipe / "recipe" / "recipe.yaml").write_text(
        rich_recipe.read_text(), encoding="utf8"
    )

    rattler_metadata = MetaData(feedstock_dir_with_recipe)

    assert rattler_metadata.name() == "rich"
    assert rattler_metadata.version() == "13.4.2"


def test_metadata_for_multiple_output(feedstock_dir_with_recipe: Path, mamba_recipe: Path) -> None:
    (feedstock_dir_with_recipe / "recipe" / "recipe.yaml").write_text(
        mamba_recipe.read_text(), encoding="utf8"
    )

    rattler_metadata = MetaData(feedstock_dir_with_recipe)

    assert rattler_metadata.name() == "mamba-split"
    assert rattler_metadata.version() == "1.5.8"


def test_metadata_when_rendering_single_output(
    feedstock_dir_with_recipe: Path, rich_recipe: Path
) -> None:
    recipe_path = feedstock_dir_with_recipe / "recipe" / "recipe.yaml"
    (recipe_path).write_text(rich_recipe.read_text(), encoding="utf8")

    rendered = render(str(recipe_path), platform="linux", arch="64")

    assert rendered[0][0].name() == "rich"
    assert rendered[0][0].version() == "13.4.2"


def test_metadata_when_rendering_multiple_output(
    feedstock_dir_with_recipe: Path, multiple_outputs: Path
) -> None:
    recipe_path = feedstock_dir_with_recipe / "recipe" / "recipe.yaml"
    (recipe_path).write_text(multiple_outputs.read_text(), encoding="utf8")

    rendered = render(str(recipe_path), platform="linux", arch="64")

    assert rendered[0][0].name() == "libmamba"
    assert rendered[0][0].version() == "1.5.8"
