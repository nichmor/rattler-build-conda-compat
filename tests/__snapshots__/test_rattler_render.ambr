# serializer version: 1
# name: test_environ_is_passed_to_rattler_build
  list([
    dict({
      'about': dict({
      }),
      'build': dict({
        'skip': list([
          "env.get('TEST_SHOULD_BE_PASSED') == 'false'",
        ]),
      }),
      'extra': dict({
        'final': True,
      }),
      'package': dict({
        'name': 'py-test',
        'version': '1.0.0',
      }),
      'requirements': dict({
        'build': list([
          'c_compiler_stub',
          'cuda_compiler_stub',
        ]),
        'host': list([
          'python',
        ]),
        'run': list([
          'python',
        ]),
      }),
    }),
  ])
# ---
# name: test_environ_is_passed_to_rattler_build.1
  dict({
    'target_platform': 'linux-64',
  })
# ---
# name: test_render_recipe
  list([
    dict({
      'python': '3.8.* *_cpython',
      'target_platform': 'linux-64',
    }),
    dict({
      'python': '3.9.* *_cpython',
      'target_platform': 'linux-64',
    }),
  ])
# ---
